theano/gof/utils.py

```diff
@@ -210,6 +211,47 @@ def __str__(self):
 
         return type.__new__(cls, name, bases, dct)
 
+    _cache = {}
+
+    def __call__(cls, *args, **kwargs):
+        _set_cache = cls.__dict__.get('_set_cache', None)
+        if _set_cache:
+            key = [cls]
+            for arg in args:
+                arg = make_hashable(arg)
+                key.append(arg)
+            for k, v in sorted(kwargs.items()):
+                key.append(k)
+                v = make_hashable(v)
```

Comment:
I don't like this idea of make_hashable, but I'm not sure we should remove it.

@lamblin @abergeron what do you think?